core/core.go

```diff
@@ -62,20 +62,20 @@ type baseSchema struct {
 }
 
 func (r *baseSchema) Solve(ctx context.Context, st llb.State, marshalOpts ...llb.ConstraintsOpt) (*filesystem.Filesystem, error) {
-	def, err := st.Marshal(ctx, append([]llb.ConstraintsOpt{llb.Platform(r.platform)}, marshalOpts...)...)
+	info, err := filesystem.ImageFromState(ctx, st, append([]llb.ConstraintsOpt{llb.Platform(r.platform)}, marshalOpts...)...)
 	if err != nil {
 		return nil, err
 	}
+
 	_, err = r.gw.Solve(ctx, bkgw.SolveRequest{
 		Evaluate:   true,
-		Definition: def.ToPB(),
+		Definition: info.FS,
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	// FIXME: should we create a filesystem from `res.SingleRef()`?
```

Comment:
Rambling/thinking/learning out loud, posting for posterity:

**tl;dr: I'll refactor the code a bit to respect this `FIXME` since we're using `SingleRef` everywhere else and it should also shrink `FSID` a bit.**

I originally removed this `FIXME` because when I tried going from `res.SingleRef` -> `ref.ToState` -> `filesystem.FromState` the image config would be lost from the original `st`.

But in hindsight that's an orthogonal problem not made any more or less possible with my change. We're just using `llb.State` as a kind of brittle vessel for this config, so I'd just need to pluck the env/dir from the input `llb.State` instead of the returned `llb.State`.

I'm not sure what `SingleRef` is actually for so I'm doing a bit of investigation first. From what I gather it seems to be a way to refer to a previously built value instead of its source LLB definition. Is this an optimization or does it have any observable behavior?

I did a gross hack to get this to use the config from the input `llb.State` and the filesystem from the output `llb.State` (it'll be less gross after I refactor) and all the tests pass. I also made it log when the marshalled value is different, and indeed it's a few kilobytes smaller sometimes, but it's also pretty frequently the same.

```patch
diff --git a/core/core.go b/core/core.go
index 758e375a..bef7402c 100644
--- a/core/core.go
+++ b/core/core.go
@@ -1,7 +1,9 @@
 package core
 
 import (
+	"bytes"
 	"context"
+	"log"
 
 	bkclient "github.com/moby/buildkit/client"
 	"github.com/moby/buildkit/client/llb"
@@ -62,20 +64,62 @@ type baseSchema struct {
 }
 
 func (r *baseSchema) Solve(ctx context.Context, st llb.State, marshalOpts ...llb.ConstraintsOpt) (*filesystem.Filesystem, error) {
-	info, err := filesystem.ImageFromState(ctx, st, append([]llb.ConstraintsOpt{llb.Platform(r.platform)}, marshalOpts...)...)
+	env, err := st.Env(ctx)
 	if err != nil {
 		return nil, err
 	}
+	dir, err := st.GetDir(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	input, err := st.Marshal(ctx, append([]llb.ConstraintsOpt{llb.Platform(r.platform)}, marshalOpts...)...)
+	if err != nil {
+		return nil, err
+	}
+
+	ibytes, _ := input.ToPB().Marshal()
 
-	_, err = r.gw.Solve(ctx, bkgw.SolveRequest{
+	res, err := r.gw.Solve(ctx, bkgw.SolveRequest{
 		Evaluate:   true,
-		Definition: info.FS,
+		Definition: input.ToPB(),
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	return info.ToFilesystem()
+	bkref, err := res.SingleRef()
+	if err != nil {
+		return nil, err
+	}
+
+	ost, err := bkref.ToState()
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := ost.Marshal(ctx, append([]llb.ConstraintsOpt{llb.Platform(r.platform)}, marshalOpts...)...)
+	if err != nil {
+		return nil, err
+	}
+
+	obytes, _ := output.ToPB().Marshal()
+
+	if bytes.Equal(ibytes, obytes) {
+		log.Println("!!! INPUT/OUTPUT SAME", len(ibytes), len(obytes))
+	} else {
+		dir, _ := os.MkdirTemp("/tmp/results", "result*")
+		os.WriteFile(filepath.Join(dir, "input"), ibytes, 0644)
+		os.WriteFile(filepath.Join(dir, "output"), obytes, 0644)
+		fmt.Println(string(ibytes), string(obytes))
+		log.Println("!!! INPUT/OUTPUT DIFF", len(ibytes), len(obytes))
+	}
+
+	return (&filesystem.Image{
+		FS: output.ToPB(),
+		Config: specs.ImageConfig{
+			Env:        env,
+			WorkingDir: dir,
+		},
+	}).ToFilesystem()
+	// return filesystem.FromState(ctx, st, r.platform)
+	// return info.ToFilesystem()
 }
 
 func (r *baseSchema) Export(ctx context.Context, fs *filesystem.Filesystem, export bkclient.ExportEntry) error {
```

Some samples from `go test -v ./...`:

```
2022/09/26 13:07:02 !!! INPUT/OUTPUT DIFF 9012 6674
2022/09/26 13:07:02 !!! INPUT/OUTPUT SAME 7269 7269
2022/09/26 13:07:03 !!! INPUT/OUTPUT SAME 7278 7278
2022/09/26 13:07:03 !!! INPUT/OUTPUT SAME 914 914
2022/09/26 13:07:03 !!! INPUT/OUTPUT SAME 550 550
2022/09/26 13:07:03 !!! INPUT/OUTPUT SAME 914 914
2022/09/26 13:07:04 !!! INPUT/OUTPUT SAME 550 550
2022/09/26 13:07:05 !!! INPUT/OUTPUT DIFF 13311 9966
2022/09/26 13:07:05 !!! INPUT/OUTPUT SAME 914 914
2022/09/26 13:07:05 !!! INPUT/OUTPUT SAME 557 557
```

The difference in the marshaled protobuf is always that the shim `Dockerfile` and `main.go` are removed:

![image](https://user-images.githubusercontent.com/1880/192341263-8c5636cc-6b93-458c-86e0-fcb79130fb1d.png)

I'm not 100% sure what the rules are for what gets omitted in the returned result, but yeah it's a bit smaller.

I'll refactor the code a bit to respect this comment since we're using it everywhere else. It should also shrink `FSID` a bit.